### DISCUSSION TITLE
WireGuard mode breaks when allowing hosts

### DISCUSSION THREAD

**Discussion by mkfmnn** [mitmproxy/mitmproxy#5930] [Q&A]
I'm using mitmproxy for the first time to inspect network traffic for an app on my iPhone. I was able to get up and running pretty quickly using the "HTTP Proxy" setting, which showed me a lot of traffic but not specifically what I was looking for; after that, I switched to WireGuard mode, and was able to find what I was looking for in a TLS-protected, non-HTTPS TCP connection. Awesome! So first of all I just wanted to express gratitude and give kudos for such a quick and easy process.

There's a lot of other traffic I don't care about capturing, so now I'm trying to get the `--allow-hosts`/`--ignore-hosts` options to work. Once I use those, though, it seems to break all traffic (iOS Safari tries to load for a while before showing a message that "the server stopped responding"). In the `mitmweb` log I only see messages related to `10.0.0.53:53` and no other hosts, so my guess is that something about those options is breaking DNS. I've tried `--allow-hosts :53` to no apparent effect, though.

Steps to reproduce:

1. Run `mitmweb --mode wireguard --allow-hosts example` on computer
2. Enable WireGuard VPN on target device (i.e. iPhone)
3. Navigate to https://example.com/ or http://example.com/ in Safari on iPhone, or any other website; observe that it fails to load.
4. Remove `--allow-hosts` argument from `mitmweb` and observe that traffic works again.

Am I doing something wrong? Is this a known issue? Any other troubleshooting I should try?

Based on my reading of the docs for how wireguard/transparent mode works, it seems like I should be able to load the page whether I'm using `--allow-hosts` or `--ignore-hosts` and whether my traffic matches the argument or not; the only difference should be whether the traffic is intercepted and logged by mitmproxy — is that understanding correct?

I'm using the mitmproxy installed via HomeBrew:

```
$ mitmproxy --version
Mitmproxy: 9.0.1
Python:    3.11.2
OpenSSL:   OpenSSL 1.1.1t  7 Feb 2023
Platform:  macOS-12.6.2-x86_64-i386-64bit
```

**Comment by mhils** [comment]
Does adding 10.0.0.53 to allow hosts help?

**Reply by mkfmnn** [reply]
It does help, but only partially.

First a couple of asides:

1. AFAICT the ignore/allow/tcp arguments only work with hostnames/IPs and not the `host:port` string documented in the [Ignoring Domains Howto](https://docs.mitmproxy.org/stable/howto-ignoredomains/), so I'm guessing that's just a documentation error with that page (maybe outdated?)—I noticed that other places like `mitmproxy --help` don't mention it matching against port at all.

2. I've been trying different combinations of command-line arguments for a while before noticing that some configurations that previously worked would stop working, or vice versa. I *think* this has something to do with my WireGuard client not connecting properly to the server in some cases, for whatever reason. I haven't recently been able to reproduce the issue of an innocuous `--ignore-hosts` "breaking everything" so I edited the title of this discussion from "ignoring" to "allowing". (`--ignore-hosts` seems to work the way I expect it to, per the documentation.)

So when I do specify only `--allow-hosts 10.0.0.53`, my client behaves normally; I can navigate to HTTP and HTTPS websites in the browser, and I see the DNS flows in the mitmweb interface. I see the `server connect` messages in the log for the websites I am navigating to, though those flows don't show up in mitmweb (as expected). Interestingly, I no longer see any `server connect 10.0.0.53:53` lines. So far so good. But there are two issues with this:

1. I shouldn't have to do that, right? Like how the HTTP/HTTPS traffic for all the websites I'm visiting is getting passed through without getting intercepted, despite not being specifically allowed. Why isn't the DNS traffic treated the same way?

2. While allowing 10.0.0.53 unblocks traffic, once I do that, I haven't been able to get anything *else* besides DNS records to show up in mitmweb. I've tried things like `--allow-hosts 10.0.0.53 --allow-hosts example` and `--allow-hosts '10.0.0.53|example'`, and while I can get http://example.com/ and https://example.com/ to *load* on the client, there's no flow for that traffic in the UI. So it's not actually getting me closer to my goal of only intercepting/capturing the traffic that I care about.

**Comment by dsphper** [comment]
@mhils 
Hi mhils,

I have managed to fix the issue. Would it be helpful if I submitted a pull request with the changes I made? I'd be happy to contribute to the project.

Best Regards.

**Comment by mhils** [comment]
Yes please, of course! :)
